src/rl/rl_controller/rl_controller/controller.py

```diff
@@ -0,0 +1,222 @@
+            # absolute angle
+            angle = abs(np.degrees(angle))
+            if angle_list == []:
+                angle_offset = angle
+            angle_list.append(abs(angle - angle_offset))
+
+        """ print("Angle list count: ", len(angle_list))
+
+        print(", ".join([f"{angle:.3f}" for angle in angle_list[0:4]]))
+        print(", ".join([f"{angle:.3f}" for angle in angle_list[4:8]]))
+        print(", ".join([f"{angle:.3f}" for angle in angle_list[8:12]]))
+        print(", ".join([f"{angle:.3f}" for angle in angle_list[12:16]]))
+        print(", ".join([f"{angle:.3f}" for angle in angle_list[16:20]]))
+        print(", ".join([f"{angle:.3f}" for angle in angle_list[20:24]]))
+        print(", ".join([f"{angle:.3f}" for angle in angle_list[24:28]]))
+        print(", ".join([f"{angle:.3f}" for angle in angle_list[28:32]]))
+        print(", ".join([f"{angle:.3f}" for angle in angle_list[32:36]]))
+        print(", ".join([f"{angle:.3f}" for angle in angle_list[36:40]])) """
+
+        self.turning_angles = angle_list
+
+    def publish_desired_speed(self, action):
+        if action == -1.0:
+            self.desired_speed = -1.0
+            msg = Float64()
+            msg.data = self.desired_speed
+            self.desired_speed_pub.publish(msg)
+            self.desired_speed = 3.0
+            return
+
+        msg = Float64()
+        self.desired_speed += 0.1 * (action - 1)
+        msg.data = self.desired_speed
+        self.desired_speed_pub.publish(msg)
+    
+    def get_observation(self):
+        # Return the current observation
+        return {
+            'angles': self.turning_angles,
+            'speed': self.desired_speed
+        }
+
+    def get_reward(self):
+        reward = self.track_point_reached
+        self.track_point_reached = 0
+        return reward
+
+    def reset_environment(self):
+        # Reset the environment to the initial state
+        print("TIMESTEPS REACHED")
+
+        self.delete_car_pub.publish(String(data="test"))
+
+        time.sleep(1)
+        self.create_car_pub.publish(String(data="test"))
+
+        time.sleep(1)
+        self.track_point_reached = 0
+        self.publish_desired_speed(-1.0)
+
+def main(args=None):
+    rclpy.init(args=args)
+    rl_environment_node = RLEnvironmentNode()
+    
+    # running Environment node in a separate thread
+    thread = threading.Thread(target=rclpy.spin, args=(rl_environment_node,))
+    thread.start()
+
+    env = CarEnv(rl_environment_node)
+
+    while (1):
+        time.sleep(0.1)
+        action = env.action_space.sample()
+        obs, reward, done, info = env.step(action)
+
+        print("  ")
+        print("Step: ")
+        print("obs: ", obs)
+        print("reward: ", reward)
+        print("done: ", done)
+        print("info: ", info)
+        print("action: ", action)
+        print("desired speed: ", rl_environment_node.desired_speed)
```

Comment:
use a logging lib so that this can be outputted into a file so that you can do training on servers